### PR TITLE
[WIP] Elasticsearch integration

### PR DESCRIPTION
**Pull request by phoenixrises**
This is the work in progress elasticsearch integration. There are still tests and documentation to write so that everything is nice, easy and most importantly that everybody can understand how the ES stuff works.

I hope to be done with these soon, but my track record is against me here.

Any issues, feel free to raise them here and I'll action them as soon possible.

### CODE REVIEW

**Comment by magicmonkey** [comment]
Thinking ahead to a live deployment, the server currently doesn't really have the memory to run ElasticSearch so we'll need to talk to our hosting provider (ie - this may not happen very quickly).  Also, from a documentation point of view, I'm interested in things like how the search documents are structured, how the index should be setup (what should be auto-analysed etc), how we manage the data in the index (is there a cron or is it done synchronously alongside updating the MySQL data) and how to initially build the index.

File: src/controllers/EventsController.php [hunk 1]
```diff
@@ -197,7 +197,29 @@ public function postAction($request, $db) {
...
                     $talk_mapper = new TalkMapper($db, $request);
-                    $new_id = $talk_mapper->save($talk);
+                    $new_id = $talk['id'] = $talk_mapper->save($talk);
+
+                    try {
+                        // Write talk to search index
+
+                        // Talks don't always have dates, but they're very useful to search.
+                        // If talk doesn't have one, try and use the event start date
+                        if(!isset($talk['start_date']) || !$talk['start_date']) {
```
Comment (akrabat) [inline src/controllers/EventsController.php]: A talk must always have a start_date, so this code is not needed.

Comment (phoenixrises) [inline src/controllers/EventsController.php]: That's great! I'll drop that code in that case.

**Comment by phoenixrises** [comment]
@magicmonkey First off, I don't think anybody is in a rush for this. It's an enhancement / replacement so in the end it's more of a nice to have when possible in my opinion.

With respect to search documents, you can see the setup of the index and it's various types and analysers in scripts/reindex-all.php which needs run in the first instance to populate existing data into ES.

From there, there is already some code that tries to generate / update documents synchronously when events / talks are updated / added. However, some functionality is missing for doing adding and editing items in the API that is done on web1 currently (as far as I can tell), so there is a limited number of places to actually get the documents updating, but as things get implemented on web2 they'll need to be updated accordingly. I had considered using a cron instead, but after thought and some discussions in the IRC channel, it just didn't make sense.

Cheers!

**Comment by magicmonkey** [comment]
Elasticsearch is installed on the server, I've been running it for a bit and watching the memory usage and it'll be OK.
I think to progress this it'd be useful to initially launch something which we can use to just populate ES with data, and then launch something which switches over to using it.  We can then stagger these, with confidence in the middle that it's all working OK in ES.

**Comment by magicmonkey** [comment]
Some feedback from a quick IRL conversation with @lornajane and @akrabat : it would be more useful to not index talk_speaker directly, and also to index users (by username and retrieve user ID) so that someone can search for a user and get a list of their comments, talks, attendances etc.

**Comment by magicmonkey** [comment]
Some feedback from another quick IRL conversation with @lornajane and @akrabat : @akrabat wants to look at this inside the dev VM to get his around the data structures a little more.

**Comment by akrabat** [comment]
The next step for this PR is for the Elasticsearch engine to be put in to the VM so that I can play with it locally.

i.e. We're looking for a PR against https://github.com/joindin/joindin-vm for this.

**Comment by phoenixrises** [comment]
@akrabat Done ;) https://github.com/joindin/joindin-vm/pull/80

**Comment by akrabat** [comment]
Okay, I've started looking at this now...

Are you still up for working on it to get it merged?

**Comment by phoenixrises** [comment]
Yep.

**Comment by PurpleBooth** [comment]
```shell
vagrant@dev:/vagrant/joindin-api$ phpcs --standard=tools/codesniffer/JoindInPSR2/ --ignore=**/config.php --ignore=**/database.php **/*.php

FILE: /vagrant/joindin-api/scripts/password-migration.php
----------------------------------------------------------------------
FOUND 3 ERRORS AFFECTING 3 LINES
----------------------------------------------------------------------
  3 | ERROR | [x] TRUE, FALSE and NULL must be lowercase; expected
    |       |     "null" but found "NULL"
 13 | ERROR | [x] Expected 1 space after WHILE keyword; 0 found
 19 | ERROR | [x] Expected 1 blank line at end of file; 2 found
----------------------------------------------------------------------
PHPCBF CAN FIX THE 3 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: /vagrant/joindin-api/scripts/patchdb.php
----------------------------------------------------------------------
FOUND 6 ERRORS AND 1 WARNING AFFECTING 7 LINES
----------------------------------------------------------------------
  20 | ERROR   | [x] Expected 1 newline after opening brace; 2 found
  88 | ERROR   | [x] Whitespace found at end of line
  96 | ERROR   | [x] Opening parenthesis of a multi-line function
     |         |     call must be the last content on the line
 104 | WARNING | [x] Usage of ELSE IF is discouraged; use ELSEIF
     |         |     instead
 117 | ERROR   | [x] Expected 1 newline after opening brace; 2 found
 140 | ERROR   | [x] Expected 1 newline after opening brace; 2 found
 162 | ERROR   | [x] Expected 1 blank line at end of file; 8 found
----------------------------------------------------------------------
PHPCBF CAN FIX THE 7 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: /vagrant/joindin-api/scripts/reindex-all.php
----------------------------------------------------------------------
FOUND 12 ERRORS AND 3 WARNINGS AFFECTING 12 LINES
----------------------------------------------------------------------
   1 | WARNING | [ ] A file should declare new symbols (classes,
     |         |     functions, constants, etc.) and cause no other
     |         |     side effects, or it should execute logic with
     |         |     side effects, but should not do both. The first
     |         |     symbol is defined on line 3 and the first side
     |         |     effect is on line 5.
  25 | ERROR   | [x] Expected 1 space after CATCH keyword; 0 found
 131 | ERROR   | [x] Whitespace found at end of line
 153 | WARNING | [ ] Line exceeds 120 characters; contains 132
     |         |     characters
 159 | ERROR   | [x] Expected 1 space after WHILE keyword; 0 found
 178 | ERROR   | [x] Expected 1 space after IF keyword; 0 found
 178 | ERROR   | [x] Whitespace found at end of line
 187 | WARNING | [ ] Line exceeds 120 characters; contains 147
     |         |     characters
 199 | ERROR   | [x] Expected 1 space after WHILE keyword; 0 found
 215 | ERROR   | [x] Expected 1 space after IF keyword; 0 found
 215 | ERROR   | [x] Whitespace found at end of line
 231 | ERROR   | [x] Expected 1 space after WHILE keyword; 0 found
 244 | ERROR   | [x] Expected 1 space after IF keyword; 0 found
 244 | ERROR   | [x] Whitespace found at end of line
 250 | ERROR   | [x] Expected 1 blank line at end of file; 6 found
----------------------------------------------------------------------
PHPCBF CAN FIX THE 12 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------

Time: 129ms; Memory: 6.25Mb
```

There's some code style issues.

File: composer.json [hunk 2]
```diff
@@ -3,6 +3,7 @@
     "require" : {
         "swiftmailer/swiftmailer": "v5.0.3",
         "michelf/php-markdown": "1.4.0",
+	    "elasticsearch/elasticsearch": "~1.0",
```
Comment (PurpleBooth) [inline composer.json]: Should be spaces not tab :)

Comment (PurpleBooth) [inline composer.json]: The composer.lock file needs updating for this library too.

**Comment by PurpleBooth** [comment]
Lots of tests failing

```shell
vagrant@dev:/vagrant/joindin-api/tests/frisby$ node_modules/.bin/jasmine-node --nocolor --junitreport --output ../../build/logs/frisby/ api_spec.js
.FF.F.........FFFFFFFFF

Failures:

  1) Frisby Test: Event search by title 
	[ GET http://api.dev.joind.in/v2.1/events/?resultsperpage=3&title=php ]
   Message:
     Expected 500 to equal 200.
   Stacktrace:
     Error: Expected 500 to equal 200.
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:493:42)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1074:43)
    at Timer.listOnTimeout (timers.js:92:15)

  2) Frisby Test: Event search by title 
	[ GET http://api.dev.joind.in/v2.1/events/?resultsperpage=3&title=php ]
   Message:
     TypeError: Cannot read property 'count' of undefined
   Stacktrace:
     TypeError: Cannot read property 'count' of undefined
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1320:35)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:212:26)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)

  3) Frisby Test: Event search by title 
	[ GET http://api.dev.joind.in/v2.1/events/?resultsperpage=3&title=php ]
   Message:
     Error: Expected 'undefined' to be type 'array' on key 'events'
   Stacktrace:
     Error: Expected 'undefined' to be type 'array' on key 'events'
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1327:15)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:213:21)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)

  4) Frisby Test: Event search by title 
	[ GET http://api.dev.joind.in/v2.1/events/?resultsperpage=3&title=php ]
   Message:
     TypeError: Cannot read property 'length' of undefined
   Stacktrace:
     TypeError: Cannot read property 'length' of undefined
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:216:26)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)
    at Timer.listOnTimeout (timers.js:92:15)

  5) Frisby Test: Event search by nonexisting title 
	[ GET http://api.dev.joind.in/v2.1/events/?title=some_nonexisting_title ]
   Message:
     Expected 500 to equal 200.
   Stacktrace:
     Error: Expected 500 to equal 200.
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:493:42)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1074:43)
    at Timer.listOnTimeout (timers.js:92:15)

  6) Frisby Test: Event search by nonexisting title 
	[ GET http://api.dev.joind.in/v2.1/events/?title=some_nonexisting_title ]
   Message:
     TypeError: Cannot read property 'count' of undefined
   Stacktrace:
     TypeError: Cannot read property 'count' of undefined
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1320:35)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:233:26)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)

  7) Frisby Test: Event search by nonexisting title 
	[ GET http://api.dev.joind.in/v2.1/events/?title=some_nonexisting_title ]
   Message:
     Error: Expected 'undefined' to be type 'array' on key 'events'
   Stacktrace:
     Error: Expected 'undefined' to be type 'array' on key 'events'
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1327:15)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:234:21)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)

  8) Frisby Test: Event search by nonexisting title 
	[ GET http://api.dev.joind.in/v2.1/events/?title=some_nonexisting_title ]
   Message:
     TypeError: Cannot read property 'length' of undefined
   Stacktrace:
     TypeError: Cannot read property 'length' of undefined
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:237:26)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)
    at Timer.listOnTimeout (timers.js:92:15)

  9) Frisby Test: Non-existent event 
	[ GET http://api.dev.joind.in/v2.1/events/100100 ]
   Message:
     Expected 500 to equal 404.
   Stacktrace:
     Error: Expected 500 to equal 404.
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:493:42)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1074:43)
    at Timer.listOnTimeout (timers.js:92:15)

  10) Frisby Test: Event list for events 
	[ GET http://api.dev.joind.in/v2.1/events ]
   Message:
     Expected 500 to equal 200.
   Stacktrace:
     Error: Expected 500 to equal 200.
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:493:42)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1074:43)
    at Timer.listOnTimeout (timers.js:92:15)

  11) Frisby Test: Event list for events 
	[ GET http://api.dev.joind.in/v2.1/events ]
   Message:
     TypeError: Cannot read property 'count' of undefined
   Stacktrace:
     TypeError: Cannot read property 'count' of undefined
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1320:35)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:45:23)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)

  12) Frisby Test: Event list for events 
	[ GET http://api.dev.joind.in/v2.1/events ]
   Message:
     Error: Expected 'undefined' to be type 'array' on key 'events'
   Stacktrace:
     Error: Expected 'undefined' to be type 'array' on key 'events'
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1327:15)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:46:18)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)

  13) Frisby Test: Event list for hot-events 
	[ GET http://api.dev.joind.in/v2.1/events?filter=hot ]
   Message:
     Expected 500 to equal 200.
   Stacktrace:
     Error: Expected 500 to equal 200.
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:493:42)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1074:43)
    at Timer.listOnTimeout (timers.js:92:15)

  14) Frisby Test: Event list for hot-events 
	[ GET http://api.dev.joind.in/v2.1/events?filter=hot ]
   Message:
     TypeError: Cannot read property 'count' of undefined
   Stacktrace:
     TypeError: Cannot read property 'count' of undefined
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1320:35)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:45:23)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)

  15) Frisby Test: Event list for hot-events 
	[ GET http://api.dev.joind.in/v2.1/events?filter=hot ]
   Message:
     Error: Expected 'undefined' to be type 'array' on key 'events'
   Stacktrace:
     Error: Expected 'undefined' to be type 'array' on key 'events'
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1327:15)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:46:18)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)

  16) Frisby Test: Event list for upcoming-events 
	[ GET http://api.dev.joind.in/v2.1/events?filter=upcoming ]
   Message:
     Expected 500 to equal 200.
   Stacktrace:
     Error: Expected 500 to equal 200.
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:493:42)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1074:43)
    at Timer.listOnTimeout (timers.js:92:15)

  17) Frisby Test: Event list for upcoming-events 
	[ GET http://api.dev.joind.in/v2.1/events?filter=upcoming ]
   Message:
     TypeError: Cannot read property 'count' of undefined
   Stacktrace:
     TypeError: Cannot read property 'count' of undefined
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1320:35)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:45:23)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)

  18) Frisby Test: Event list for upcoming-events 
	[ GET http://api.dev.joind.in/v2.1/events?filter=upcoming ]
   Message:
     Error: Expected 'undefined' to be type 'array' on key 'events'
   Stacktrace:
     Error: Expected 'undefined' to be type 'array' on key 'events'
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1327:15)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:46:18)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)

  19) Frisby Test: Event list for past-events 
	[ GET http://api.dev.joind.in/v2.1/events?filter=past ]
   Message:
     Expected 500 to equal 200.
   Stacktrace:
     Error: Expected 500 to equal 200.
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:493:42)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1074:43)
    at Timer.listOnTimeout (timers.js:92:15)

  20) Frisby Test: Event list for past-events 
	[ GET http://api.dev.joind.in/v2.1/events?filter=past ]
   Message:
     TypeError: Cannot read property 'count' of undefined
   Stacktrace:
     TypeError: Cannot read property 'count' of undefined
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1320:35)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:45:23)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)

  21) Frisby Test: Event list for past-events 
	[ GET http://api.dev.joind.in/v2.1/events?filter=past ]
   Message:
     Error: Expected 'undefined' to be type 'array' on key 'events'
   Stacktrace:
     Error: Expected 'undefined' to be type 'array' on key 'events'
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1327:15)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:46:18)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)

  22) Frisby Test: Event list for open-cfps 
	[ GET http://api.dev.joind.in/v2.1/events?filter=cfp ]
   Message:
     Expected 500 to equal 200.
   Stacktrace:
     Error: Expected 500 to equal 200.
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:493:42)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1074:43)
    at Timer.listOnTimeout (timers.js:92:15)

  23) Frisby Test: Event list for open-cfps 
	[ GET http://api.dev.joind.in/v2.1/events?filter=cfp ]
   Message:
     TypeError: Cannot read property 'count' of undefined
   Stacktrace:
     TypeError: Cannot read property 'count' of undefined
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1320:35)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:45:23)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)
    at Timer.listOnTimeout (timers.js:92:15)

  24) Frisby Test: Event list for open-cfps 
	[ GET http://api.dev.joind.in/v2.1/events?filter=cfp ]
   Message:
     Error: Expected 'undefined' to be type 'array' on key 'events'
   Stacktrace:
     Error: Expected 'undefined' to be type 'array' on key 'events'
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1327:15)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:46:18)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)
    at Timer.listOnTimeout (timers.js:92:15)

  25) Frisby Test: Event list for search 
	[ GET http://api.dev.joind.in/v2.1/search ]
   Message:
     Error: Expected 'undefined' to be type 'array' on key 'events'
   Stacktrace:
     Error: Expected 'undefined' to be type 'array' on key 'events'
    at _jsonContainsTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1327:15)
    at jasmine.Matchers.toContainJsonTypes (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1201:12)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:46:18)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)

  26) Frisby Test: Event detail for SymfonyCongress 
	[ GET http://api.dev.joind.in/v2.1/events/45?verbose=yes ]
   Message:
     Expected 500 to equal 200.
   Stacktrace:
     Error: Expected 500 to equal 200.
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:493:42)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1074:43)
    at Timer.listOnTimeout (timers.js:92:15)

  27) Frisby Test: Event detail for SymfonyCongress 
	[ GET http://api.dev.joind.in/v2.1/events/45?verbose=yes ]
   Message:
     TypeError: Cannot read property '0' of undefined
   Stacktrace:
     TypeError: Cannot read property '0' of undefined
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:66:29)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)
    at Timer.listOnTimeout (timers.js:92:15)

  28) Frisby Test: Event detail for PHPCon 
	[ GET http://api.dev.joind.in/v2.1/events/5?verbose=yes ]
   Message:
     Expected 500 to equal 200.
   Stacktrace:
     Error: Expected 500 to equal 200.
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:493:42)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1074:43)
    at Timer.listOnTimeout (timers.js:92:15)

  29) Frisby Test: Event detail for PHPCon 
	[ GET http://api.dev.joind.in/v2.1/events/5?verbose=yes ]
   Message:
     TypeError: Cannot read property '0' of undefined
   Stacktrace:
     TypeError: Cannot read property '0' of undefined
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:66:29)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)
    at Timer.listOnTimeout (timers.js:92:15)

  30) Frisby Test: Event detail for MySQL day 
	[ GET http://api.dev.joind.in/v2.1/events/33?verbose=yes ]
   Message:
     Expected 500 to equal 200.
   Stacktrace:
     Error: Expected 500 to equal 200.
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:493:42)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1074:43)
    at Timer.listOnTimeout (timers.js:92:15)

  31) Frisby Test: Event detail for MySQL day 
	[ GET http://api.dev.joind.in/v2.1/events/33?verbose=yes ]
   Message:
     TypeError: Cannot read property '0' of undefined
   Stacktrace:
     TypeError: Cannot read property '0' of undefined
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/api_read.js:66:29)
    at Frisby.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:951:8)
    at null.<anonymous> (/vagrant/joindin-api/tests/frisby/node_modules/frisby/lib/frisby.js:1112:18)
    at Timer.listOnTimeout (timers.js:92:15)

Finished in 0.737 seconds
23 tests, 362 assertions, 31 failures, 0 skipped

```

File: src/services/SearchService.php [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+    /**
+     * Remove document from the search index in the provided type
+     * 
+     * @param string $type Define the index's type we want to remove from
+     * @param int $id The ID of the document we want to remove
+     *
+     * @return bool
+     */
+    public function remove($type, $id) {
```
Comment (PurpleBooth) [inline src/services/SearchService.php]: It's worth splitting these classes into separate services

This is a very long class as it stands

File: src/services/SearchService.php [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        $params['body']['query']['function_score']['score_mode'] = 'multiply';
+        $params['body']['query']['function_score']['boost_mode'] = 'multiply';
+
+        $results = $this->provider->search($params);
+
+        return $results;
+    }
+
+    public function write($index, $data) {
+        switch($index) {
```
Comment (PurpleBooth) [inline src/services/SearchService.php]: Change this to use polymorphism to avoid massive switches

**Comment by PurpleBooth** [comment]
The actual ElasticSearch stuff looks good to me, the mapping is explicitly set, dates are dates. Looks good!

**Comment by lornajane** [comment]
I would still like this feature but I don't see much value in keeping this PR open right now as I think there's more to do before we can accept - I'm closing this PR but I'm still very much +1 on the feature!

**Pull request closed without merging**